rfcs/2021-09-21-9292-throttle-transform.md

```diff
@@ -0,0 +1,149 @@
+# RFC 9292 - 2021-09-21 - `throttle` transform
+
+This RFC proposes the addition of a new transform that provides a user the ability to control the throughput of specific event streams.
+
+## Context
+
+- [#258](https://github.com/vectordotdev/vector/issues/258)
+
+## Scope
+
+### In scope
+
+- Dropping events (logs and metrics) to rate limit an event stream
+- Rate limit by both bytes (unserialized) and events
+- Exclude events with VRL conditions
+- Specify buckets based on the value of a key in the event
+
+### Out of scope
+
+- Apply backpressure as a form of rate limiting
+
+## Pain
+
+- Users lack the necessary tooling to control throughput which can cause excessive costs and negatively impact downstream services due to increased load
+- Admins cannot set quotas for users/usergroups utilizing Vector
+
+## Proposal
+
+### User Experience
+
+The `throttle` transform can be used to rate limit specific subsets of your event stream to limit load on downstream services or to enforce quotas on users.
+You can enforce rate limits on events or their raw size, as well as excluding events based on a VRL condition to avoid dropping critical logs. Rate limits
```

Comment:
> as well as excluding events based on a VRL condition to avoid dropping critical logs.

I can see the convenience of having this, but it's worth pointing you can combine this and the `route` transform to achieve the same thing:

```
event -> route -> (non-throttled logs) ----+---> sink
           |                               |
           +----> throttle transform ------+
```

If we still want to add this feature, it's worth expanding on why we think it's worth adding this "shortcut" here, but not (yet) in other transforms (e.g. `sample`).

Reply:
👍 I pulled in the "exclude" option here based on [`sample`](https://vector.dev/docs/reference/configuration/transforms/sample/#exclude)'s existing configuration option. I can see both ways being reasonable, but I lean towards this existing configuration strategy.

Reply:
Ah, d'oh. I totally missed that we already have this in the `sample` transform.